In [1]:
import os
os.path

<module 'posixpath' from '/home/tylerf/tensorflow/lib/python3.5/posixpath.py'>

In [4]:
import logging
import os
import pickle
import pandas
from multiprocessing import Pool
import math
import numpy as np
import matplotlib.pyplot as plt
from datetime import datetime
import keras
import csv

Using TensorFlow backend.


In [130]:
class linear():
    """
    Linear Model Class: A SLR model class
    param: theta - slope for 
    param: sigma - Standard deviation
    param: n - sample size
    function: simulate() - simulate model data
    function: plot() - plot model data with true
    TODO: make lin_x and lin_y attributes?
    """

    def __init__(self, theta=2, sigma=10, x_min=0, x_max=20):
        self.sigma = sigma
        self.theta = theta
        self.x_min = x_min
        self.x_max = x_max
        self.x_dim = 1
        self.y_dim = 1

    def simulate(self, n):
        x = np.sort(np.random.uniform(self.x_min, self.x_max, n))
        y = self.theta * x + np.random.normal(0, self.sigma, n)
        return pandas.DataFrame({'x': x, 'y': y})

    def plot(self, sample_space):
        lin_x = np.linspace(self.x_min, self.x_max, 100)
        lin_y = self.theta * lin_x
        matplotlib.pyplot.scatter(sample_space['x'], sample_space['y'])
        matplotlib.pyplot.plot(lin_x, lin_y)
        matplotlib.pyplot.show()


class nonlinear():
    """
    Linear Model Class: A SLR model class
    param: theta - slope for
    param: sigma - Standard deviation
    param: n - sample size
    simulate - function
    TODO: make lin_x and lin_y attributes?
    """

    def __init__(self, n=1000, sigma=10, theta_1=80, theta_2=71, theta_3=-10, theta_4=2.4, x_min=0, x_max=90):
        self.n = n
        self.sigma = sigma
        self.theta_1 = 80
        self.theta_2 = 71
        self.theta_3 = -10
        self.theta_4 = 2.4
        self.x_min = 0
        self.x_max = 90
        #self.df = self.simulate()

    def simulate(self,sample_size):
        x = sorted(np.random.uniform(self.x_min, self.x_max, sample_size))
        y = self.theta_1 - self.theta_2 * np.exp(
        -np.exp(self.theta_3 + self.theta_4 * np.log(x))) + np.random.normal(0, self.sigma, sample_size)
        return pandas.DataFrame({'x': x, 'y': y})

    def theta(self, x=45):
        return(self.theta_1 - self.theta_2 * np.exp(
        -np.exp(self.theta_3 + self.theta_4 * np.log(x))))

    def plot(self):
        lin_x = np.linspace(self.x_min, self.x_max, 100)
        lin_y = self.theta_1 - self.theta_2 * np.exp(-np.exp(self.theta_3 + self.theta_4 * np.log(lin_x)))
        matplotlib.pyplot.scatter(self.df['x'], self.df['y'])
        matplotlib.pyplot.plot(lin_x, lin_y)
        matplotlib.pyplot.show()


class mars():
    def __init__(self, sample_size=1000):
        self.x_min = 0
        self.x_max = 1
        self.x_dim = 5
        self.sample_size = sample_size
        self.y_dim = 1
        

    def simulate(self):
        x = np.random.uniform(self.x_min, self.x_max, size=(self.sample_size, 5))     
        y = 10*np.sin(np.pi*x[:,0]*x[:,1]) - 20*(x[:,2]-0.05)**2 + 10*x[:,3] + 5*x[:,4] + np.random.normal(0, 10, self.sample_size)
        return (x,y)


In [126]:
# s=np.random.uniform(0, 1, size = (3,5))
print(s)
s[:,1]

[[0.37070824 0.813501   0.75511687 0.25120105 0.0779309 ]
 [0.72681793 0.41334843 0.26066188 0.20718389 0.50728908]
 [0.59961194 0.23996353 0.66394911 0.05422526 0.30008556]]


array([0.813501  , 0.41334843, 0.23996353])

In [137]:
trainX, trainY= mars(100000).simulate()
trainX.shape

x_pred = np.array([.5,.5,.5,.5,.5])
x_pred.shape

(5,)

In [161]:
debug = False
num_lay_nodes = 12
lr = .01
epochs = 3000
keras_verbose = 1
trainX, trainY = mars(10000).simulate()

def build_mars_estimator():  

    nn1 = keras.Sequential()
    # This is worth about 9% if we cut neurons down to 25
    nn1.add(keras.layers.Dense(num_lay_nodes, kernel_initializer="uniform", 
                               activation='sigmoid', input_dim=5))
    nn1.add(keras.layers.Dense(units=1, activation='linear'))
    # training
    # sgd = keras.optimizers.SGD(lr=0.0001);
    adam = keras.optimizers.Adam(lr=lr)
    # model.compile(loss='mse', optimizer=sgd)
    nn1.compile(loss='mse', optimizer=adam)
    return(nn1)


    
def estimate(nn,trainX, trainY):    
    
    
    ##This isn't strictly necessary but maintaisn the same intialiation 
    ## parameterization for each neural network. 
    Wsave = nn.get_weights()
    nn.fit(trainX, trainY, epochs=epochs,verbose=keras_verbose)

    # predictions
    # Linear predict 10 
    #x_pred = np.arange(10, 11, 1, dtype='float32').reshape(-1, 1)
   
    x_pred = np.array([.5,.5,.5,.5,.5])
    predictions = nn.predict(x_pred.reshape((1,5)))
    nn.set_weights(Wsave)
    
    if debug:
        plt.scatter(simulated_data['x'], simulated_data['y'], edgecolors='g')
        plt.plot(x_pred, predictions, 'r')
        plt.legend(['Predicted Y', 'Actual Y'])
        plt.show()
    
    return(predictions)

estimator = build_mars_estimator()
p = list()
start_time = datetime.now()
for _ in range(1):    
    p.append(estimate(estimator, trainX, trainY))

print("This took {} seconds to run".format(datetime.now()-start_time))
print("Mean {} Var {}".format(np.mean(p),np.var(p)))




Epoch 1/3000
10000/10000 [==============================] - 1s 70us/step - loss: 145.2676
Epoch 2/3000
10000/10000 [==============================] - 0s 48us/step - loss: 114.9400
Epoch 3/3000
10000/10000 [==============================] - 0s 50us/step - loss: 109.5012
Epoch 4/3000
10000/10000 [==============================] - 0s 50us/step - loss: 108.6767
Epoch 5/3000
10000/10000 [==============================] - 0s 48us/step - loss: 108.4974
Epoch 6/3000
10000/10000 [==============================] - 0s 49us/step - loss: 108.4600
Epoch 7/3000
10000/10000 [==============================] - 1s 50us/step - loss: 108.3904
Epoch 8/3000
10000/10000 [==============================] - 0s 50us/step - loss: 108.2596
Epoch 9/3000
10000/10000 [==============================] - 1s 52us/step - loss: 108.1386
Epoch 10/3000
10000/10000 [==============================] - 1s 51us/step - loss: 108.1967
Epoch 11/3000
10000/10000 [==============================] - 1s 54us/step - loss: 108.1149
Epoch 12

10000/10000 [==============================] - 0s 47us/step - loss: 103.2764
Epoch 92/3000
10000/10000 [==============================] - 0s 49us/step - loss: 103.3733
Epoch 93/3000
10000/10000 [==============================] - 0s 47us/step - loss: 103.2824
Epoch 94/3000
10000/10000 [==============================] - 0s 47us/step - loss: 103.0668
Epoch 95/3000
10000/10000 [==============================] - 0s 48us/step - loss: 103.0921
Epoch 96/3000
10000/10000 [==============================] - 0s 47us/step - loss: 103.1271
Epoch 97/3000
10000/10000 [==============================] - 0s 47us/step - loss: 103.1251
Epoch 98/3000
10000/10000 [==============================] - 0s 46us/step - loss: 103.0877
Epoch 99/3000
10000/10000 [==============================] - 0s 47us/step - loss: 103.1143
Epoch 100/3000
10000/10000 [==============================] - 0s 47us/step - loss: 103.0120
Epoch 101/3000
10000/10000 [==============================] - 0s 49us/step - loss: 102.9982
Epoch 102/3

10000/10000 [==============================] - 0s 48us/step - loss: 101.7993
Epoch 181/3000
10000/10000 [==============================] - 0s 48us/step - loss: 101.7227
Epoch 182/3000
10000/10000 [==============================] - 0s 48us/step - loss: 101.6675
Epoch 183/3000
10000/10000 [==============================] - 0s 47us/step - loss: 101.7358
Epoch 184/3000
10000/10000 [==============================] - 0s 47us/step - loss: 101.7374
Epoch 185/3000
10000/10000 [==============================] - 0s 47us/step - loss: 101.7449
Epoch 186/3000
10000/10000 [==============================] - 0s 47us/step - loss: 101.7753
Epoch 187/3000
10000/10000 [==============================] - 0s 48us/step - loss: 101.7205
Epoch 188/3000
10000/10000 [==============================] - 0s 47us/step - loss: 101.5808
Epoch 189/3000
10000/10000 [==============================] - 0s 48us/step - loss: 101.5900
Epoch 190/3000
10000/10000 [==============================] - 1s 55us/step - loss: 101.7191
Epo

10000/10000 [==============================] - 0s 47us/step - loss: 101.3751
Epoch 270/3000
10000/10000 [==============================] - 0s 48us/step - loss: 101.2399
Epoch 271/3000
10000/10000 [==============================] - 0s 46us/step - loss: 101.3919
Epoch 272/3000
10000/10000 [==============================] - 0s 47us/step - loss: 101.3337
Epoch 273/3000
10000/10000 [==============================] - 0s 46us/step - loss: 101.4006
Epoch 274/3000
10000/10000 [==============================] - 0s 47us/step - loss: 101.3916
Epoch 275/3000
10000/10000 [==============================] - 0s 47us/step - loss: 101.3966
Epoch 276/3000
10000/10000 [==============================] - 0s 48us/step - loss: 101.4064
Epoch 277/3000
10000/10000 [==============================] - 0s 47us/step - loss: 101.3918
Epoch 278/3000
10000/10000 [==============================] - 0s 49us/step - loss: 101.3495
Epoch 279/3000
10000/10000 [==============================] - 0s 48us/step - loss: 101.3381
Epo

10000/10000 [==============================] - 0s 47us/step - loss: 101.1226
Epoch 359/3000
10000/10000 [==============================] - 0s 48us/step - loss: 101.2236
Epoch 360/3000
10000/10000 [==============================] - 0s 47us/step - loss: 101.1643
Epoch 361/3000
10000/10000 [==============================] - 0s 47us/step - loss: 101.2274
Epoch 362/3000
10000/10000 [==============================] - 0s 48us/step - loss: 101.1004
Epoch 363/3000
10000/10000 [==============================] - 0s 47us/step - loss: 101.1039
Epoch 364/3000
10000/10000 [==============================] - 0s 49us/step - loss: 101.1139
Epoch 365/3000
10000/10000 [==============================] - 0s 47us/step - loss: 101.2041
Epoch 366/3000
10000/10000 [==============================] - 1s 51us/step - loss: 101.1094
Epoch 367/3000
10000/10000 [==============================] - 1s 50us/step - loss: 101.2291
Epoch 368/3000
10000/10000 [==============================] - 1s 53us/step - loss: 101.1689
Epo

10000/10000 [==============================] - 1s 50us/step - loss: 100.9886
Epoch 448/3000
10000/10000 [==============================] - 0s 48us/step - loss: 100.9356
Epoch 449/3000
10000/10000 [==============================] - 0s 47us/step - loss: 101.1496
Epoch 450/3000
10000/10000 [==============================] - 0s 46us/step - loss: 100.9096
Epoch 451/3000
10000/10000 [==============================] - 0s 48us/step - loss: 101.0515
Epoch 452/3000
10000/10000 [==============================] - 0s 47us/step - loss: 101.0091
Epoch 453/3000
10000/10000 [==============================] - 0s 48us/step - loss: 101.0103
Epoch 454/3000
10000/10000 [==============================] - 0s 46us/step - loss: 101.0486
Epoch 455/3000
10000/10000 [==============================] - 0s 49us/step - loss: 100.9769
Epoch 456/3000
10000/10000 [==============================] - 0s 46us/step - loss: 100.9807
Epoch 457/3000
10000/10000 [==============================] - 0s 50us/step - loss: 101.0037
Epo

10000/10000 [==============================] - 0s 48us/step - loss: 100.9168
Epoch 537/3000
10000/10000 [==============================] - 0s 48us/step - loss: 101.0164
Epoch 538/3000
10000/10000 [==============================] - 1s 52us/step - loss: 100.8460
Epoch 539/3000
10000/10000 [==============================] - 0s 47us/step - loss: 100.9810
Epoch 540/3000
10000/10000 [==============================] - 0s 48us/step - loss: 100.9512
Epoch 541/3000
10000/10000 [==============================] - 0s 47us/step - loss: 100.8750
Epoch 542/3000
10000/10000 [==============================] - 1s 51us/step - loss: 100.8236
Epoch 543/3000
10000/10000 [==============================] - 0s 48us/step - loss: 100.9096
Epoch 544/3000
10000/10000 [==============================] - 0s 48us/step - loss: 100.9107
Epoch 545/3000
10000/10000 [==============================] - 0s 48us/step - loss: 100.8760
Epoch 546/3000
10000/10000 [==============================] - 0s 48us/step - loss: 100.8919
Epo

10000/10000 [==============================] - 0s 49us/step - loss: 100.8467
Epoch 626/3000
10000/10000 [==============================] - 0s 49us/step - loss: 100.7424
Epoch 627/3000
10000/10000 [==============================] - 0s 49us/step - loss: 100.7299
Epoch 628/3000
10000/10000 [==============================] - 0s 48us/step - loss: 100.7556
Epoch 629/3000
10000/10000 [==============================] - 0s 46us/step - loss: 100.7942
Epoch 630/3000
10000/10000 [==============================] - 0s 49us/step - loss: 100.7799
Epoch 631/3000
10000/10000 [==============================] - 0s 47us/step - loss: 100.7618
Epoch 632/3000
10000/10000 [==============================] - 0s 47us/step - loss: 100.7877
Epoch 633/3000
10000/10000 [==============================] - 0s 47us/step - loss: 100.7897
Epoch 634/3000
10000/10000 [==============================] - 0s 48us/step - loss: 100.8038
Epoch 635/3000
10000/10000 [==============================] - 0s 49us/step - loss: 100.8257
Epo

10000/10000 [==============================] - 0s 47us/step - loss: 100.7781
Epoch 715/3000
10000/10000 [==============================] - 0s 48us/step - loss: 100.7473
Epoch 716/3000
10000/10000 [==============================] - 0s 47us/step - loss: 100.7611
Epoch 717/3000
10000/10000 [==============================] - 0s 48us/step - loss: 100.7239
Epoch 718/3000
10000/10000 [==============================] - 0s 47us/step - loss: 100.7128
Epoch 719/3000
10000/10000 [==============================] - 0s 47us/step - loss: 100.7291
Epoch 720/3000
10000/10000 [==============================] - 0s 48us/step - loss: 100.7006
Epoch 721/3000
10000/10000 [==============================] - 0s 47us/step - loss: 100.8390
Epoch 722/3000
10000/10000 [==============================] - 0s 48us/step - loss: 100.6979
Epoch 723/3000
10000/10000 [==============================] - 1s 50us/step - loss: 100.7116
Epoch 724/3000
10000/10000 [==============================] - 0s 47us/step - loss: 100.6795
Epo

10000/10000 [==============================] - 0s 48us/step - loss: 100.6256
Epoch 804/3000
10000/10000 [==============================] - 0s 49us/step - loss: 100.6720
Epoch 805/3000
10000/10000 [==============================] - 0s 48us/step - loss: 100.6011
Epoch 806/3000
10000/10000 [==============================] - 0s 48us/step - loss: 100.6591
Epoch 807/3000
10000/10000 [==============================] - 0s 49us/step - loss: 100.5789
Epoch 808/3000
10000/10000 [==============================] - 0s 47us/step - loss: 100.5913
Epoch 809/3000
10000/10000 [==============================] - 0s 48us/step - loss: 100.6516
Epoch 810/3000
10000/10000 [==============================] - 0s 46us/step - loss: 100.6909
Epoch 811/3000
10000/10000 [==============================] - 0s 47us/step - loss: 100.5944
Epoch 812/3000
10000/10000 [==============================] - 0s 47us/step - loss: 100.7742
Epoch 813/3000
10000/10000 [==============================] - 0s 48us/step - loss: 100.6394
Epo

10000/10000 [==============================] - 0s 47us/step - loss: 100.5546
Epoch 893/3000
10000/10000 [==============================] - 0s 48us/step - loss: 100.6238
Epoch 894/3000
10000/10000 [==============================] - 0s 46us/step - loss: 100.5351
Epoch 895/3000
10000/10000 [==============================] - 0s 49us/step - loss: 100.6634
Epoch 896/3000
10000/10000 [==============================] - 0s 47us/step - loss: 100.6178
Epoch 897/3000
10000/10000 [==============================] - 0s 49us/step - loss: 100.6177
Epoch 898/3000
10000/10000 [==============================] - 0s 47us/step - loss: 100.6129
Epoch 899/3000
10000/10000 [==============================] - 0s 48us/step - loss: 100.6271
Epoch 900/3000
10000/10000 [==============================] - 0s 46us/step - loss: 100.7137
Epoch 901/3000
10000/10000 [==============================] - 0s 48us/step - loss: 100.6146
Epoch 902/3000
10000/10000 [==============================] - 0s 48us/step - loss: 100.6310
Epo

10000/10000 [==============================] - 0s 48us/step - loss: 100.5631
Epoch 982/3000
10000/10000 [==============================] - 0s 49us/step - loss: 100.5597
Epoch 983/3000
10000/10000 [==============================] - 0s 47us/step - loss: 100.5698
Epoch 984/3000
10000/10000 [==============================] - 0s 49us/step - loss: 100.6759
Epoch 985/3000
10000/10000 [==============================] - 0s 47us/step - loss: 100.6281
Epoch 986/3000
10000/10000 [==============================] - 0s 47us/step - loss: 100.6236
Epoch 987/3000
10000/10000 [==============================] - 0s 48us/step - loss: 100.5433
Epoch 988/3000
10000/10000 [==============================] - 0s 47us/step - loss: 100.5701
Epoch 989/3000
10000/10000 [==============================] - 0s 50us/step - loss: 100.4848
Epoch 990/3000
10000/10000 [==============================] - 0s 50us/step - loss: 100.6657
Epoch 991/3000
10000/10000 [==============================] - 0s 47us/step - loss: 100.5585
Epo

10000/10000 [==============================] - 0s 48us/step - loss: 100.5584
Epoch 1070/3000
10000/10000 [==============================] - 0s 47us/step - loss: 100.5243
Epoch 1071/3000
10000/10000 [==============================] - 0s 47us/step - loss: 100.5756
Epoch 1072/3000
10000/10000 [==============================] - 0s 48us/step - loss: 100.6365
Epoch 1073/3000
10000/10000 [==============================] - 0s 47us/step - loss: 100.5672
Epoch 1074/3000
10000/10000 [==============================] - 0s 48us/step - loss: 100.5776
Epoch 1075/3000
10000/10000 [==============================] - 0s 48us/step - loss: 100.6312
Epoch 1076/3000
10000/10000 [==============================] - 0s 49us/step - loss: 100.5475
Epoch 1077/3000
10000/10000 [==============================] - 0s 46us/step - loss: 100.6646
Epoch 1078/3000
10000/10000 [==============================] - 1s 51us/step - loss: 100.5468
Epoch 1079/3000
10000/10000 [==============================] - 0s 48us/step - loss: 10

10000/10000 [==============================] - 0s 48us/step - loss: 100.5744
Epoch 1158/3000
10000/10000 [==============================] - 0s 49us/step - loss: 100.5938
Epoch 1159/3000
10000/10000 [==============================] - 0s 48us/step - loss: 100.5378
Epoch 1160/3000
10000/10000 [==============================] - 0s 48us/step - loss: 100.6258
Epoch 1161/3000
10000/10000 [==============================] - 0s 46us/step - loss: 100.5928
Epoch 1162/3000
10000/10000 [==============================] - 0s 47us/step - loss: 100.5577
Epoch 1163/3000
10000/10000 [==============================] - 0s 49us/step - loss: 100.6011
Epoch 1164/3000
10000/10000 [==============================] - 0s 48us/step - loss: 100.6406
Epoch 1165/3000
10000/10000 [==============================] - 0s 48us/step - loss: 100.5647
Epoch 1166/3000
10000/10000 [==============================] - 0s 49us/step - loss: 100.6055
Epoch 1167/3000
10000/10000 [==============================] - 0s 48us/step - loss: 10

10000/10000 [==============================] - 0s 48us/step - loss: 100.5698
Epoch 1246/3000
10000/10000 [==============================] - 0s 48us/step - loss: 100.5025
Epoch 1247/3000
10000/10000 [==============================] - 0s 47us/step - loss: 100.6127
Epoch 1248/3000
10000/10000 [==============================] - 0s 48us/step - loss: 100.5431
Epoch 1249/3000
10000/10000 [==============================] - 0s 46us/step - loss: 100.5524
Epoch 1250/3000
10000/10000 [==============================] - 0s 49us/step - loss: 100.5951
Epoch 1251/3000
10000/10000 [==============================] - 0s 48us/step - loss: 100.6105
Epoch 1252/3000
10000/10000 [==============================] - 0s 48us/step - loss: 100.6510
Epoch 1253/3000
10000/10000 [==============================] - 0s 47us/step - loss: 100.5512
Epoch 1254/3000
10000/10000 [==============================] - 0s 48us/step - loss: 100.5737
Epoch 1255/3000
10000/10000 [==============================] - 0s 46us/step - loss: 10

10000/10000 [==============================] - 0s 48us/step - loss: 100.5735
Epoch 1334/3000
10000/10000 [==============================] - 0s 48us/step - loss: 100.6414
Epoch 1335/3000
10000/10000 [==============================] - 0s 49us/step - loss: 100.5507
Epoch 1336/3000
10000/10000 [==============================] - 0s 46us/step - loss: 100.5312
Epoch 1337/3000
10000/10000 [==============================] - 0s 50us/step - loss: 100.4762
Epoch 1338/3000
10000/10000 [==============================] - 0s 49us/step - loss: 100.5650
Epoch 1339/3000
10000/10000 [==============================] - 0s 49us/step - loss: 100.6157
Epoch 1340/3000
10000/10000 [==============================] - 0s 46us/step - loss: 100.5114
Epoch 1341/3000
10000/10000 [==============================] - 0s 47us/step - loss: 100.5191
Epoch 1342/3000
10000/10000 [==============================] - 0s 47us/step - loss: 100.5000
Epoch 1343/3000
10000/10000 [==============================] - 0s 47us/step - loss: 10

10000/10000 [==============================] - 0s 46us/step - loss: 100.4743
Epoch 1422/3000
10000/10000 [==============================] - 0s 48us/step - loss: 100.5707
Epoch 1423/3000
10000/10000 [==============================] - 0s 48us/step - loss: 100.5290
Epoch 1424/3000
10000/10000 [==============================] - 1s 51us/step - loss: 100.6013
Epoch 1425/3000
10000/10000 [==============================] - 0s 49us/step - loss: 100.5512
Epoch 1426/3000
10000/10000 [==============================] - 0s 48us/step - loss: 100.5336
Epoch 1427/3000
10000/10000 [==============================] - 0s 47us/step - loss: 100.5891
Epoch 1428/3000
10000/10000 [==============================] - 0s 46us/step - loss: 100.5007
Epoch 1429/3000
10000/10000 [==============================] - 0s 48us/step - loss: 100.5565
Epoch 1430/3000
10000/10000 [==============================] - 0s 46us/step - loss: 100.5590
Epoch 1431/3000
10000/10000 [==============================] - 0s 47us/step - loss: 10

10000/10000 [==============================] - 0s 47us/step - loss: 100.5711
Epoch 1510/3000
10000/10000 [==============================] - 0s 47us/step - loss: 100.5567
Epoch 1511/3000
10000/10000 [==============================] - 0s 47us/step - loss: 100.5477
Epoch 1512/3000
10000/10000 [==============================] - 0s 47us/step - loss: 100.5869
Epoch 1513/3000
10000/10000 [==============================] - 0s 49us/step - loss: 100.5229
Epoch 1514/3000
10000/10000 [==============================] - 0s 49us/step - loss: 100.4112
Epoch 1515/3000
10000/10000 [==============================] - 0s 49us/step - loss: 100.5276
Epoch 1516/3000
10000/10000 [==============================] - 0s 48us/step - loss: 100.5343
Epoch 1517/3000
10000/10000 [==============================] - 0s 48us/step - loss: 100.5735
Epoch 1518/3000
10000/10000 [==============================] - 0s 48us/step - loss: 100.4785
Epoch 1519/3000
10000/10000 [==============================] - 0s 47us/step - loss: 10

10000/10000 [==============================] - 0s 47us/step - loss: 100.5031
Epoch 1598/3000
10000/10000 [==============================] - 0s 48us/step - loss: 100.6147
Epoch 1599/3000
10000/10000 [==============================] - 0s 48us/step - loss: 100.4496
Epoch 1600/3000
10000/10000 [==============================] - 0s 47us/step - loss: 100.5455
Epoch 1601/3000
10000/10000 [==============================] - 0s 49us/step - loss: 100.5250
Epoch 1602/3000
10000/10000 [==============================] - 0s 49us/step - loss: 100.5112
Epoch 1603/3000
10000/10000 [==============================] - 0s 49us/step - loss: 100.4801
Epoch 1604/3000
10000/10000 [==============================] - 0s 46us/step - loss: 100.5289
Epoch 1605/3000
10000/10000 [==============================] - 0s 48us/step - loss: 100.5274
Epoch 1606/3000
10000/10000 [==============================] - 1s 52us/step - loss: 100.5234
Epoch 1607/3000
10000/10000 [==============================] - 1s 51us/step - loss: 10

10000/10000 [==============================] - 0s 49us/step - loss: 100.4880
Epoch 1686/3000
10000/10000 [==============================] - 0s 50us/step - loss: 100.6546
Epoch 1687/3000
10000/10000 [==============================] - 0s 46us/step - loss: 100.4715
Epoch 1688/3000
10000/10000 [==============================] - 0s 48us/step - loss: 100.5828
Epoch 1689/3000
10000/10000 [==============================] - 0s 47us/step - loss: 100.4303
Epoch 1690/3000
10000/10000 [==============================] - 1s 58us/step - loss: 100.5132
Epoch 1691/3000
10000/10000 [==============================] - 1s 50us/step - loss: 100.5789
Epoch 1692/3000
10000/10000 [==============================] - 1s 50us/step - loss: 100.5549
Epoch 1693/3000
10000/10000 [==============================] - 1s 50us/step - loss: 100.5322
Epoch 1694/3000
10000/10000 [==============================] - 1s 51us/step - loss: 100.4846
Epoch 1695/3000
10000/10000 [==============================] - 0s 48us/step - loss: 10

10000/10000 [==============================] - 0s 48us/step - loss: 100.5026
Epoch 1774/3000
10000/10000 [==============================] - 0s 47us/step - loss: 100.5117
Epoch 1775/3000
10000/10000 [==============================] - 0s 47us/step - loss: 100.5909
Epoch 1776/3000
10000/10000 [==============================] - 0s 47us/step - loss: 100.4918
Epoch 1777/3000
10000/10000 [==============================] - 0s 46us/step - loss: 100.5630
Epoch 1778/3000
10000/10000 [==============================] - 0s 48us/step - loss: 100.4855
Epoch 1779/3000
10000/10000 [==============================] - 0s 46us/step - loss: 100.5013
Epoch 1780/3000
10000/10000 [==============================] - 0s 47us/step - loss: 100.6022
Epoch 1781/3000
10000/10000 [==============================] - 0s 47us/step - loss: 100.5674
Epoch 1782/3000
10000/10000 [==============================] - 0s 48us/step - loss: 100.5133
Epoch 1783/3000
10000/10000 [==============================] - 0s 46us/step - loss: 10

10000/10000 [==============================] - 0s 48us/step - loss: 100.5817
Epoch 1862/3000
10000/10000 [==============================] - 0s 47us/step - loss: 100.4723
Epoch 1863/3000
10000/10000 [==============================] - 0s 47us/step - loss: 100.5530
Epoch 1864/3000
10000/10000 [==============================] - 0s 47us/step - loss: 100.5797
Epoch 1865/3000
10000/10000 [==============================] - 1s 51us/step - loss: 100.5382
Epoch 1866/3000
10000/10000 [==============================] - 0s 47us/step - loss: 100.5572
Epoch 1867/3000
10000/10000 [==============================] - 0s 47us/step - loss: 100.5881
Epoch 1868/3000
10000/10000 [==============================] - 0s 48us/step - loss: 100.5313
Epoch 1869/3000
10000/10000 [==============================] - 0s 47us/step - loss: 100.5632
Epoch 1870/3000
10000/10000 [==============================] - 0s 47us/step - loss: 100.5391
Epoch 1871/3000
10000/10000 [==============================] - 0s 48us/step - loss: 10

10000/10000 [==============================] - 0s 47us/step - loss: 100.5806
Epoch 1950/3000
10000/10000 [==============================] - 0s 46us/step - loss: 100.4648
Epoch 1951/3000
10000/10000 [==============================] - 0s 47us/step - loss: 100.5114
Epoch 1952/3000
10000/10000 [==============================] - 0s 48us/step - loss: 100.4885
Epoch 1953/3000
10000/10000 [==============================] - 0s 49us/step - loss: 100.4897
Epoch 1954/3000
10000/10000 [==============================] - 0s 47us/step - loss: 100.5378
Epoch 1955/3000
10000/10000 [==============================] - 0s 48us/step - loss: 100.5386
Epoch 1956/3000
10000/10000 [==============================] - 0s 47us/step - loss: 100.5060
Epoch 1957/3000
10000/10000 [==============================] - 0s 48us/step - loss: 100.4934
Epoch 1958/3000
10000/10000 [==============================] - 0s 49us/step - loss: 100.5255
Epoch 1959/3000
10000/10000 [==============================] - 0s 50us/step - loss: 10

10000/10000 [==============================] - 0s 46us/step - loss: 100.4788
Epoch 2038/3000
10000/10000 [==============================] - 0s 48us/step - loss: 100.5248
Epoch 2039/3000
10000/10000 [==============================] - 0s 46us/step - loss: 100.6089
Epoch 2040/3000
10000/10000 [==============================] - 0s 49us/step - loss: 100.4918
Epoch 2041/3000
10000/10000 [==============================] - 0s 46us/step - loss: 100.3926
Epoch 2042/3000
10000/10000 [==============================] - 0s 48us/step - loss: 100.5092
Epoch 2043/3000
10000/10000 [==============================] - 0s 47us/step - loss: 100.4969
Epoch 2044/3000
10000/10000 [==============================] - 0s 48us/step - loss: 100.5180
Epoch 2045/3000
10000/10000 [==============================] - 0s 47us/step - loss: 100.4496
Epoch 2046/3000
10000/10000 [==============================] - 0s 46us/step - loss: 100.5037
Epoch 2047/3000
10000/10000 [==============================] - 0s 49us/step - loss: 10

10000/10000 [==============================] - 0s 48us/step - loss: 100.4965
Epoch 2126/3000
10000/10000 [==============================] - 0s 48us/step - loss: 100.5637
Epoch 2127/3000
10000/10000 [==============================] - 0s 46us/step - loss: 100.4683
Epoch 2128/3000
10000/10000 [==============================] - 0s 49us/step - loss: 100.4103
Epoch 2129/3000
10000/10000 [==============================] - 0s 47us/step - loss: 100.5655
Epoch 2130/3000
10000/10000 [==============================] - 0s 47us/step - loss: 100.4854
Epoch 2131/3000
10000/10000 [==============================] - 0s 50us/step - loss: 100.4448
Epoch 2132/3000
10000/10000 [==============================] - 1s 50us/step - loss: 100.6577
Epoch 2133/3000
10000/10000 [==============================] - 0s 48us/step - loss: 100.5701
Epoch 2134/3000
10000/10000 [==============================] - 0s 49us/step - loss: 100.4819
Epoch 2135/3000
10000/10000 [==============================] - 0s 46us/step - loss: 10

10000/10000 [==============================] - 0s 50us/step - loss: 100.5467
Epoch 2214/3000
10000/10000 [==============================] - 0s 48us/step - loss: 100.4915
Epoch 2215/3000
10000/10000 [==============================] - 0s 47us/step - loss: 100.5425
Epoch 2216/3000
10000/10000 [==============================] - 0s 48us/step - loss: 100.3983
Epoch 2217/3000
10000/10000 [==============================] - 0s 48us/step - loss: 100.5026
Epoch 2218/3000
10000/10000 [==============================] - 0s 48us/step - loss: 100.4675
Epoch 2219/3000
10000/10000 [==============================] - 0s 48us/step - loss: 100.5447
Epoch 2220/3000
10000/10000 [==============================] - 0s 50us/step - loss: 100.5569
Epoch 2221/3000
10000/10000 [==============================] - 0s 49us/step - loss: 100.5072
Epoch 2222/3000
10000/10000 [==============================] - 0s 48us/step - loss: 100.4334
Epoch 2223/3000
10000/10000 [==============================] - 0s 49us/step - loss: 10

10000/10000 [==============================] - 0s 47us/step - loss: 100.4340
Epoch 2302/3000
10000/10000 [==============================] - 0s 48us/step - loss: 100.5081
Epoch 2303/3000
10000/10000 [==============================] - 0s 46us/step - loss: 100.5351
Epoch 2304/3000
10000/10000 [==============================] - 0s 49us/step - loss: 100.4732
Epoch 2305/3000
10000/10000 [==============================] - 0s 48us/step - loss: 100.5615
Epoch 2306/3000
10000/10000 [==============================] - 0s 49us/step - loss: 100.4764
Epoch 2307/3000
10000/10000 [==============================] - 0s 46us/step - loss: 100.4852
Epoch 2308/3000
10000/10000 [==============================] - 0s 48us/step - loss: 100.5591
Epoch 2309/3000
10000/10000 [==============================] - 0s 48us/step - loss: 100.5216
Epoch 2310/3000
10000/10000 [==============================] - 0s 49us/step - loss: 100.4344
Epoch 2311/3000
10000/10000 [==============================] - 0s 47us/step - loss: 10

10000/10000 [==============================] - 0s 47us/step - loss: 100.5755
Epoch 2390/3000
10000/10000 [==============================] - 0s 49us/step - loss: 100.4548
Epoch 2391/3000
10000/10000 [==============================] - 0s 47us/step - loss: 100.4678
Epoch 2392/3000
10000/10000 [==============================] - 0s 48us/step - loss: 100.5237
Epoch 2393/3000
10000/10000 [==============================] - 0s 47us/step - loss: 100.5655
Epoch 2394/3000
10000/10000 [==============================] - 0s 48us/step - loss: 100.4834
Epoch 2395/3000
10000/10000 [==============================] - 0s 50us/step - loss: 100.4629
Epoch 2396/3000
10000/10000 [==============================] - 0s 48us/step - loss: 100.5193
Epoch 2397/3000
10000/10000 [==============================] - 0s 47us/step - loss: 100.5874
Epoch 2398/3000
10000/10000 [==============================] - 0s 48us/step - loss: 100.4787
Epoch 2399/3000
10000/10000 [==============================] - 0s 48us/step - loss: 10

10000/10000 [==============================] - 0s 48us/step - loss: 100.4563
Epoch 2478/3000
10000/10000 [==============================] - 0s 47us/step - loss: 100.4104
Epoch 2479/3000
10000/10000 [==============================] - 0s 48us/step - loss: 100.5486
Epoch 2480/3000
10000/10000 [==============================] - 0s 47us/step - loss: 100.4516
Epoch 2481/3000
10000/10000 [==============================] - 0s 48us/step - loss: 100.6088
Epoch 2482/3000
10000/10000 [==============================] - 0s 47us/step - loss: 100.5682
Epoch 2483/3000
10000/10000 [==============================] - 0s 48us/step - loss: 100.5485
Epoch 2484/3000
10000/10000 [==============================] - 0s 47us/step - loss: 100.4603
Epoch 2485/3000
10000/10000 [==============================] - 0s 49us/step - loss: 100.5096
Epoch 2486/3000
10000/10000 [==============================] - 0s 47us/step - loss: 100.5323
Epoch 2487/3000
10000/10000 [==============================] - 0s 47us/step - loss: 10

10000/10000 [==============================] - 0s 47us/step - loss: 100.4956
Epoch 2566/3000
10000/10000 [==============================] - 0s 49us/step - loss: 100.4668
Epoch 2567/3000
10000/10000 [==============================] - 0s 47us/step - loss: 100.5101
Epoch 2568/3000
10000/10000 [==============================] - 0s 50us/step - loss: 100.4756
Epoch 2569/3000
10000/10000 [==============================] - 0s 47us/step - loss: 100.4538
Epoch 2570/3000
10000/10000 [==============================] - 0s 47us/step - loss: 100.4549
Epoch 2571/3000
10000/10000 [==============================] - 0s 47us/step - loss: 100.5087
Epoch 2572/3000
10000/10000 [==============================] - 0s 47us/step - loss: 100.5662
Epoch 2573/3000
10000/10000 [==============================] - 0s 49us/step - loss: 100.4383
Epoch 2574/3000
10000/10000 [==============================] - 0s 47us/step - loss: 100.4955
Epoch 2575/3000
10000/10000 [==============================] - 0s 47us/step - loss: 10

10000/10000 [==============================] - 0s 47us/step - loss: 100.4964
Epoch 2654/3000
10000/10000 [==============================] - 0s 47us/step - loss: 100.4908
Epoch 2655/3000
10000/10000 [==============================] - 0s 46us/step - loss: 100.5057
Epoch 2656/3000
10000/10000 [==============================] - 0s 47us/step - loss: 100.5608
Epoch 2657/3000
10000/10000 [==============================] - 0s 47us/step - loss: 100.4747
Epoch 2658/3000
10000/10000 [==============================] - 0s 48us/step - loss: 100.5354
Epoch 2659/3000
10000/10000 [==============================] - 0s 46us/step - loss: 100.4710
Epoch 2660/3000
10000/10000 [==============================] - 0s 48us/step - loss: 100.5321
Epoch 2661/3000
10000/10000 [==============================] - 0s 47us/step - loss: 100.5171
Epoch 2662/3000
10000/10000 [==============================] - 0s 47us/step - loss: 100.5085
Epoch 2663/3000
10000/10000 [==============================] - 0s 48us/step - loss: 10

10000/10000 [==============================] - 0s 48us/step - loss: 100.6073
Epoch 2742/3000
10000/10000 [==============================] - 0s 46us/step - loss: 100.4980
Epoch 2743/3000
10000/10000 [==============================] - 0s 48us/step - loss: 100.5465
Epoch 2744/3000
10000/10000 [==============================] - 0s 47us/step - loss: 100.4882
Epoch 2745/3000
10000/10000 [==============================] - 0s 47us/step - loss: 100.4726
Epoch 2746/3000
10000/10000 [==============================] - 0s 48us/step - loss: 100.4716
Epoch 2747/3000
10000/10000 [==============================] - 1s 50us/step - loss: 100.5353
Epoch 2748/3000
10000/10000 [==============================] - 0s 47us/step - loss: 100.5101
Epoch 2749/3000
10000/10000 [==============================] - 0s 47us/step - loss: 100.5612
Epoch 2750/3000
10000/10000 [==============================] - 0s 47us/step - loss: 100.4720
Epoch 2751/3000
10000/10000 [==============================] - 0s 47us/step - loss: 10

10000/10000 [==============================] - 0s 48us/step - loss: 100.5116
Epoch 2830/3000
10000/10000 [==============================] - 0s 48us/step - loss: 100.4831
Epoch 2831/3000
10000/10000 [==============================] - 0s 48us/step - loss: 100.4802
Epoch 2832/3000
10000/10000 [==============================] - 0s 47us/step - loss: 100.4842
Epoch 2833/3000
10000/10000 [==============================] - 0s 49us/step - loss: 100.5152
Epoch 2834/3000
10000/10000 [==============================] - 0s 48us/step - loss: 100.4664
Epoch 2835/3000
10000/10000 [==============================] - 0s 47us/step - loss: 100.4520
Epoch 2836/3000
10000/10000 [==============================] - 0s 49us/step - loss: 100.5060
Epoch 2837/3000
10000/10000 [==============================] - 0s 49us/step - loss: 100.4045
Epoch 2838/3000
10000/10000 [==============================] - 0s 47us/step - loss: 100.4626
Epoch 2839/3000
10000/10000 [==============================] - 0s 47us/step - loss: 10

10000/10000 [==============================] - 0s 48us/step - loss: 100.5613
Epoch 2918/3000
10000/10000 [==============================] - 0s 48us/step - loss: 100.5026
Epoch 2919/3000
10000/10000 [==============================] - 0s 47us/step - loss: 100.4092
Epoch 2920/3000
10000/10000 [==============================] - 0s 48us/step - loss: 100.4573
Epoch 2921/3000
10000/10000 [==============================] - 0s 49us/step - loss: 100.4840
Epoch 2922/3000
10000/10000 [==============================] - 0s 49us/step - loss: 100.4351
Epoch 2923/3000
10000/10000 [==============================] - 0s 46us/step - loss: 100.4573
Epoch 2924/3000
10000/10000 [==============================] - 0s 48us/step - loss: 100.5031
Epoch 2925/3000
10000/10000 [==============================] - 0s 47us/step - loss: 100.4000
Epoch 2926/3000
10000/10000 [==============================] - 0s 48us/step - loss: 100.4011
Epoch 2927/3000
10000/10000 [==============================] - 0s 48us/step - loss: 10

AttributeError: 'list' object has no attribute 'shape'

In [149]:
estimator.fit(trainX, trainY, epochs=epochs,verbose=keras_verbose)


In [160]:
x=np.array([.5,.5,.5,.5,.5]).reshape(1,5)
print(10*np.sin(np.pi*x[:,0]*x[:,1]) - 20*(x[:,2]-0.05)**2 + 10*x[:,3] + 5*x[:,4]) 


[10.52106781]


In [35]:
x1=.5
x2=.5
x3=.5
x4=.5
x5=.5
print(10*np.sin(np.pi*x1*x2) - 20*(x3-0.05)**2 + 10*x4 + 5*x5 )


10.521067811865475


p

In [142]:
#numpy.random.seed(seed=None)

n_mc = 51
n_z = 3
k_n = 30
n=1000
# num_processes = 12
in_interval_2 = 0
in_interval_3 = 0
num_trials = 120
theta = nonlinear().theta()
start_time = datetime.now()
estimates = list()

for _ in range(0, num_trials):

    nz_average = list() 
    iv_pred = list()

    for i in range(0,n_z):
        fixed_point = random_sample.sample(1)    
        random_sample.drop(fixed_point.index)
        for _ in range(0,n_mc):
    #         with Pool(num_processes) as p:
    #             iv_pred.append(p.apply_async(estimate, 
    #                                          estimator, random_sample.sample(k_n).append(fixed_point)))
            iv_pred.append(estimate(estimator, random_sample.sample(k_n).append(fixed_point)))
        nz_average.append(np.mean(iv_pred[(i*n_mc):(i+1)*n_mc]))    
        random_sample.append(fixed_point) 

    zeta_1_k = np.var(nz_average)
    zeta_k_k = np.var(iv_pred)
    theta_k = np.mean(iv_pred)


    alpha = n / (n_mc*n_z)
    var_net = k_n**2 / alpha * zeta_1_k + zeta_k_k

   
 
    SE = (var_net / (n_mc * n_z))**(1/2)
   
    print("Zeta_1_K {}, Zeta_k_k {}, Theta_k {} SE {}".format(
        zeta_1_k, zeta_k_k, theta_k, SE))
    
    if(theta < theta_k + 2*SE  and theta > theta_k - 2*SE):
        in_interval_2 +=1
    if(theta < theta_k + 3*SE  and theta > theta_k - 3*SE):
        in_interval_3 +=1
    
    estimates.append((theta_k,SE))

print("This took {} seconds to run".format(datetime.now()-start_time))    
print("This is the coverage rate 95% {} 99.7% {}".format(
    in_interval_2 / num_trials, in_interval_3/num_trials))


Zeta_1_K 1.1576563119888306, Zeta_k_k 21.45486831665039, Theta_k 32.61699295043945 SE 1.087252766510343
Zeta_1_K 0.031117716804146767, Zeta_k_k 19.33538246154785, Theta_k 33.69559860229492 SE 0.3929134687078297
Zeta_1_K 0.005681868642568588, Zeta_k_k 22.582788467407227, Theta_k 33.732337951660156 SE 0.3907858830637161
Zeta_1_K 1.1675390005111694, Zeta_k_k 23.591114044189453, Theta_k 33.038448333740234 SE 1.0977137096395269
Zeta_1_K 0.05830392614006996, Zeta_k_k 15.775700569152832, Theta_k 34.081878662109375 SE 0.39443971348333257
Zeta_1_K 0.2346401810646057, Zeta_k_k 28.3934268951416, Theta_k 33.129520416259766 SE 0.6298842088202895
Zeta_1_K 0.4831989109516144, Zeta_k_k 23.871707916259766, Theta_k 34.00550842285156 SE 0.7687023185884434
Zeta_1_K 4.449669361114502, Zeta_k_k 29.502233505249023, Theta_k 34.232547760009766 SE 2.04878683141827
Zeta_1_K 2.0920002460479736, Zeta_k_k 23.283300399780273, Theta_k 33.801300048828125 SE 1.4265267804807131
Zeta_1_K 1.279849648475647, Zeta_k_k 26.26

Zeta_1_K 0.7372872829437256, Zeta_k_k 23.703134536743164, Theta_k 34.909637451171875 SE 0.9046993990281649
Zeta_1_K 1.609169840812683, Zeta_k_k 24.449417114257812, Theta_k 32.740699768066406 SE 1.2680902840844215
Zeta_1_K 0.6317813396453857, Zeta_k_k 30.540016174316406, Theta_k 33.31426239013672 SE 0.876476556750144
Zeta_1_K 0.2568223476409912, Zeta_k_k 24.8713321685791, Theta_k 33.778926849365234 SE 0.6274534560179841
Zeta_1_K 0.24484817683696747, Zeta_k_k 18.889497756958008, Theta_k 33.77042007446289 SE 0.5863651838288073
Zeta_1_K 0.008370445109903812, Zeta_k_k 24.07423210144043, Theta_k 33.645511627197266 SE 0.40605581245753675
Zeta_1_K 0.2350500077009201, Zeta_k_k 24.050168991088867, Theta_k 33.14414596557617 SE 0.6072360775937221
Zeta_1_K 0.8156879544258118, Zeta_k_k 21.86898422241211, Theta_k 33.92280960083008 SE 0.9365114502154349
Zeta_1_K 0.2012457698583603, Zeta_k_k 24.849008560180664, Theta_k 33.30803298950195 SE 0.5861168940811209
Zeta_1_K 0.0015861369902268052, Zeta_k_k 22.

In [126]:
theta = nonlinear().theta()
theta

33.41813569769473

In [143]:
stable_estimates_153_3000 = estimates.copy()
stable_estimates_153_3000

[(32.616993, 1.087252766510343),
 (33.6956, 0.3929134687078297),
 (33.732338, 0.3907858830637161),
 (33.03845, 1.0977137096395269),
 (34.08188, 0.39443971348333257),
 (33.12952, 0.6298842088202895),
 (34.00551, 0.7687023185884434),
 (34.232548, 2.04878683141827),
 (33.8013, 1.4265267804807131),
 (33.172985, 1.1504498898583817),
 (32.913944, 0.45533968031377997),
 (33.4081, 0.5523615131233287),
 (32.728, 0.5929344081499014),
 (33.736366, 1.1331169441851332),
 (32.29056, 0.474817715680134),
 (33.274227, 0.7258807451208761),
 (33.306995, 0.40625443545737555),
 (33.37333, 0.7988121079511362),
 (33.327026, 0.6474067930169103),
 (33.811314, 1.3894095320249107),
 (33.450207, 0.5676986365580399),
 (33.29663, 0.6162610736999818),
 (33.923702, 0.84112862617021),
 (32.80776, 0.516230706140393),
 (33.210907, 0.44986183833521326),
 (33.50256, 0.47266833950767817),
 (33.42406, 0.8121732934063762),
 (33.771206, 0.6259391177496292),
 (34.383904, 1.3901661252972235),
 (33.595066, 0.5110031736783451),
 

In [138]:
#stable_estimates.append(estimates)
stable_estimates_51_3000=stable_estimates_51_3000+estimates

In [144]:
with open('estimates_nonlinear_153_3000.csv', 'w') as f:
    writer = csv.writer(f , lineterminator='\n')
    for tup in stable_estimates_153_3000:
        writer.writerow(tup)

In [134]:
print("Zeta_1_K {}, Zeta_k_k {}, Theta_k {}".format(zeta_1_k, zeta_k_k, theta_k))

Zeta_1_K 0.7391712665557861, Zeta_k_k 18.45136260986328, Theta_k 34.3473014831543


In [65]:
fixed_point = random_sample.sample(1)
random_sample.drop(fixed_point.index)
print(fixed_point)   
print(random_sample.sample(1).append(fixed_point))   


            x         y
151  2.843675  3.636818
            x         y
403  8.277868  9.062615
151  2.843675  3.636818


# External Coverage Study

In [38]:


def external(estimator):    
    n_mc = 30
    n_z = 30
    k_n = 31
    m_n = 30
    n=1000
        
    random_sample = nonlinear().simulate(n)
    pred = list()
    
    
    for _ in range(m_n):
        pred.append(estimate(estimator, random_sample.sample(k_n)))  
    
    
    mc_pred = list()
    for _ in range(n_z):
        fixed_point = random_sample.sample(1)    
        random_sample.drop(fixed_point.index)
        iv_pred = list()
        for _ in range(0,n_mc):
            iv_pred.append(estimate(estimator, random_sample.sample(k_n-1).append(fixed_point)))
        random_sample.append(fixed_point) 
        mc_pred.append(np.mean(iv_pred))

    nz_pred = list()
    for _ in range(n_z):
        nz_pred.append(estimate(estimator, random_sample.sample(k_n)))  

    zeta_1_k = np.var(mc_pred)
    zeta_k_k = np.var(nz_pred)
    theta_k = np.mean(pred)


    alpha = n / (m_n)
    var_net = k_n**2 / alpha * zeta_1_k + zeta_k_k

   
 
    SE = (var_net / (m_n))**(1/2)
    print("Zeta_1_K {}, Zeta_k_k {}, Theta_k {} SE {}".format(
    zeta_1_k, zeta_k_k, theta_k, SE))
    return((theta_k, SE))

    
    
def coverage_prob(theta_k, SE, ci):
    theta = nonlinear().theta()
    
    if(theta < theta_k + 2*SE  and theta > theta_k - 2*SE):
        ci[0] +=1
    if(theta < theta_k + 3*SE  and theta > theta_k - 3*SE):
        ci[1] +=1
    
    return(ci)
    

    



n_mc = 30
n_z = 30
k_n = 31
m_n = 30
n=1000
# num_processes = 12
in_interval_2 = 0
in_interval_3 = 0
num_trials = 50
start_time = datetime.now()
estimates = list()

# estimator = build_linear_estimator()



# with Pool(8) as p:
#         print(p.map(external, [0]))
estimator = build_linear_estimator()
for _ in range(num_trials):       
    estimates.append(external(estimator))

print("This took {} seconds to run".format(datetime.now()-start_time))    
# print("This is the coverage rate 95% {} 99.7% {}".format(
#     in_interval_2 / num_trials, in_interval_3/num_trials))




Zeta_1_K 1.3307031393051147, Zeta_k_k 20.732728958129883, Theta_k 33.16799545288086 SE 1.4035300788166998
Zeta_1_K 1.8167887926101685, Zeta_k_k 13.05528450012207, Theta_k 34.075157165527344 SE 1.4768582124572558
Zeta_1_K 0.890783965587616, Zeta_k_k 18.113676071166992, Theta_k 32.595157623291016 SE 1.2082353219890287
Zeta_1_K 0.6629222631454468, Zeta_k_k 21.30449867248535, Theta_k 31.813154220581055 SE 1.1606973122361772
Zeta_1_K 1.2972795963287354, Zeta_k_k 17.31553840637207, Theta_k 33.58317565917969 SE 1.3505074252360296
Zeta_1_K 1.5637774467468262, Zeta_k_k 25.45545196533203, Theta_k 33.732391357421875 SE 1.533396619219818
Zeta_1_K 1.290860652923584, Zeta_k_k 16.106128692626953, Theta_k 33.14133834838867 SE 1.3331871751110054
Zeta_1_K 0.7285254001617432, Zeta_k_k 15.268172264099121, Theta_k 33.852535247802734 SE 1.099568999665508
Zeta_1_K 1.1787950992584229, Zeta_k_k 19.899606704711914, Theta_k 35.43388748168945 SE 1.3402023406477626
Zeta_1_K 1.3933953046798706, Zeta_k_k 17.17437744

In [39]:
with open('nonlinear_3000_external_30_30_50.csv', 'w') as f:
    writer = csv.writer(f , lineterminator='\n')
    for tup in estimates:
        writer.writerow(tup)


In [40]:
ci=[0,0]  
for i in range(len(estimates)):
    ci = coverage_prob(estimates[i][0], estimates[i][1], ci)
print("Coverage rate, CI2 {} CI3{}".format(ci[0]/50, ci[1]/50))

Coverage rate, CI2 0.92 CI31.0


# Linear External Coverage Study 1:
n_mc = 30
n_z = 3
k_n = 31
m_n = 30
n=1000
num_lay_nodes = 10
lr = .01
epochs = 1000
keras_verbose = 0
random_sample = linear().simulate(1000)
num_trials = 150



Zeta_k_k 32.042850494384766 N_Z 30
                            

# Linear External Coverage Study 2:
n_mc = 30
n_z = 30
k_n = 31
m_n = 30
n=1000
num_lay_nodes = 10
lr = .01
epochs = 1000
keras_verbose = 0
random_sample = linear().simulate(1000)
num_trials = 50

Results .96 CI2 and .98 CI3

# Nonlinear External Coverage Study 1:
n_mc = 30
n_z = 30
k_n = 31
m_n = 30
n=1000
num_lay_nodes = 10
lr = .01
epochs = 3000
keras_verbose = 0
random_sample = linear().simulate(1000)
num_trials = 50

Results .92 CI2 and 1 CI3